### PR TITLE
Add support for burp client/server services

### PR DESCRIPTION
**Pull request by unode**
###### Motivation for this change
BURP is currently available in nixpkgs but provides no services

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This is a reiteration of #30553 but basing on `unode/nixpkgs:burp` instead of `unode/nixpkgs:master` to avoid accidental mixing of work. Unfortunately GitHub doesn't allow editing the head branch.

~This setup is mostly functional but there is currently a problem with validity of the SSL certificates after pairing - see https://github.com/grke/burp/issues/769.~ If anyone wants to give this a try feedback is welcome.

### CODE REVIEW

**Comment by unode** [comment]
Rebased to fix conflicts with master.

Pinging @tokudan @fpletz @ryantm as potential reviewers based on maintenance of burp.

File: nixos/modules/services/backup/burp.nix [hunk 1]
```diff
@@ -0,0 +1,653 @@
...
+        description = ''
+          Location where client files are stored
+        '';
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "burp";
+        description = ''
+          Run server as user
```
Comment (tokudan) [inline nixos/modules/services/backup/burp.nix]: Is this the client or the server config?

File: nixos/modules/services/backup/burp.nix [hunk 2]
```diff
@@ -0,0 +1,653 @@
...
+        type = types.str;
+        default = "127.0.0.1";
+        description = ''
+          Address of burp server the client should contact
+        '';
+      };
+
+      cname = mkOption {
+        type = types.str;
+        default = "${config.networking.hostName}-nixos";
```
Comment (tokudan) [inline nixos/modules/services/backup/burp.nix]: Why an extra "-nixos" here?

File: nixos/modules/services/backup/burp.nix [hunk 3]
```diff
@@ -0,0 +1,653 @@
...
+    };
+
+    systemd.services.burp-server = mkIf cfg.server.enable {
+      description = "BURP Server";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      preStart = ''
+        if [ ! -f "${cfg.server.ssl_cert}" ]; then
+          mkdir -m 0750 -p ${cfg.server.libDir}
```
Comment (tokudan) [inline nixos/modules/services/backup/burp.nix]: Why allow group read access to that directory?

File: nixos/modules/services/backup/burp.nix [hunk 4]
```diff
@@ -0,0 +1,653 @@
...
+          mkdir -m 0750 -p ${cfg.client.libDir}
+          # As of burp-2.1.32 we still need to create this ourselves
+          mkdir -m 0750 -p ${cfg.client.ca_csr_dir}
+
+          # This is need with current burp stable
+          # In the future we should be able to reference ${burp_conf} directly
+          # See also https://github.com/grke/burp/issues/620
+          cp -f ${burp_conf} ${cfg.client.libDir}/burp.conf
+
+          # This may fail so keep trying until we get ${cfg.client.ssl_cert} from the server
```
[the remark this answers is not in the document]
Comment (unode) [reply nixos/modules/services/backup/burp.nix]: Any suggestion on how to achieve that? I've seen this pattern used in other services but didn't notice a throttling mechanism.

**Comment by unode** [comment]
All points but the throttling have been addressed. I'm unsure how to proceed on that front.  
By default systemd already gives up after repeated unsuccessful starts.

**Comment by tokudan** [comment]
If systemd gives up or throttles on its own after a few attempts, that's probably good enough.
I wasn't aware that systemd does that, if the service is set to restart automatically.

Looks good to me. I think it's ready for merge.

**Comment by Infinisil** [comment]
Not sure what I think of adding another 50+ options. Every options currently decreases evaluation time for all NixOS users (whether they use the options or not), lots of options aren't very maintainable (upstream defaults and options change).

As an alternative, check out what I did in https://github.com/NixOS/nixpkgs/pull/45470, and try to apply the same strategy to this PR. This can reduce the option count to a minimum while still being able to set all current and future config values conveniently, but giving up having them all documented and type checked by the module system.

**Comment by Infinisil** [comment]
> I'm guessing you mean "increases" there grinning otherwise this would actually be a good thing.

Yes thanks, fixed now :)

> I could very easily move everything non-essential to extraConfig.

That would be fine too by me

> As for the solution you linked, was that a one-off or is this a pattern that is growing in popularity?

Well, I'm very much a favor of this approach, but I've only seen it used a couple other times, it's not very popular. I'm encouraging people to use it though. I'm thinking about writing some nixpkgs tooling to make writing such config options easier, by providing standard types and converters for common config file formats like JSON/YAML/INI. I've also started writing some general guidelines for writing NixOS services which should limit the amount of options one needs to use and increase code reuse.

**Comment by pstn** [comment]
@Infinisil What pattern are you talking off exactly? Is it documented? I'm willing to put some time in to get this merged.

**Comment by tokudan** [comment]
I believe one of the issues is that burp's config looks like an INI file, but isn't. It must support multiple lines with the same key for some options like "keep". So this would need a special converter.
So the question is actually: Do we want a module that's basically ready to merge or do we reject it and ask for a rewrite of a special converter that's a lot more complicated than just having some additional options.

**Comment by Infinisil** [comment]
I've since opened an RFC for this, see NixOS/rfcs#42. I'm currently in the process of rewriting it to address the discussions.

The second part of it is that it encourages people to keep option counts low. I think it's always a good idea to start with options that make sense for most people and tell users to use a structural `config`/`settings` option with a flexible type for when they need more (that's the first part of the RFC). More options can always be added later if they're really needed, in contrast options can't really be removed once added. See also here: https://github.com/NixOS/rfcs/pull/42#issuecomment-507083505

And while yes, the RFC isn't accepted yet (although it most likely will be eventually), and the general convenience functions aren't yet written, I'm still encouraging that approach, because it just makes so much sense. Every module I merge that doesn't adhere to that approach is a module that can't really be changed to that approach ever. So I'm not really feeling comfortable merging ones like this (which use `extraConfig` and have loads of options).

**Comment by unode** [comment]
I've removed some non-essential options. Didn't yet manage to implement @Infinisil 's suggestion above.

2 checks are now failing but this seems to be unrelated, at least the previous commit fails locally as well.

**Comment by tokudan** [comment]
It's been some time, but with 20.03 split off, maybe this could be merged if it still works?

**Comment by unode** [comment]
I've been using it on my laptop without issues ever since I opened the PR.

**Comment by unode** [comment]
I'm still using the original recipe and would like to get it merged at some point.
Will revisit once 21.05 lands on my system.

File: nixos/modules/misc/ids.nix [hunk 5]
```diff
@@ -346,6 +346,7 @@ in
       zoneminder = 314;
       paperless = 315;
       #mailman = 316;  # removed 2019-08-30
+      burp = 317;
```
Comment (SuperSandro2000) [inline nixos/modules/misc/ids.nix:349]: Can we use dynamic users?

File: nixos/modules/services/backup/burp.nix [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+        description = ''
+          Controls how frequently the systemd timer is triggered and a backup
+          is attempted.
+          Backup frequency is still determined by the server, this
+          controls only how often the client tries.
+        '';
+      };
+
+      port = mkOption {
+        type = types.int;
```
Comment (SuperSandro2000) [inline nixos/modules/services/backup/burp.nix:87]:
```suggestion
        type = types.port;
```

File: nixos/modules/services/backup/burp.nix [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+    ${cfg.server.extraConfig}
+  '';
+
+in {
+  options = {
+    services.burp.client = {
+      enable = mkEnableOption "BURP client";
+
+      frequency = mkOption {
+        type = types.str;
```
Comment (SuperSandro2000) [inline nixos/modules/services/backup/burp.nix:76]: Should this be an enum?

File: nixos/modules/services/backup/burp.nix [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+      port = mkOption {
+        type = types.int;
+        default = 4971;
+        description = ''
+          Port used for backup communication
+        '';
+      };
+
+      status_port = mkOption {
+        type = types.int;
```
Comment (SuperSandro2000) [inline nixos/modules/services/backup/burp.nix:95]:
```suggestion
        type = types.port;
```

File: nixos/modules/services/backup/burp.nix [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+      addresses = mkOption {
+        type = types.listOf types.str;
+        default = [ "127.0.0.1" ];
+        description = ''
+          Addresses to listen on
+        '';
+      };
+
+      port = mkOption {
+        type = types.int;
```
Comment (SuperSandro2000) [inline nixos/modules/services/backup/burp.nix:245]:
```suggestion
        type = types.port;
```

File: nixos/modules/services/backup/burp.nix [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+        default = 4971;
+        description = ''
+          Port used for backup communication
+        '';
+      };
+
+      status_port = mkOption {
+        type = types.int;
```
Comment (SuperSandro2000) [inline nixos/modules/services/backup/burp.nix:253]:
```suggestion
        type = types.port;
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: nixos/modules/misc/ids.nix (hunk 5 above)
Comment (unode) [reply nixos/modules/misc/ids.nix:349]: I'm not familiar with this feature. Do you have a link I could read up on? Thanks!

File: nixos/modules/services/backup/burp.nix (hunk 2 above)
Comment (unode) [reply nixos/modules/services/backup/burp.nix:112]:
```suggestion
        default = "${config.networking.hostName}";
```

File: nixos/modules/services/backup/burp.nix (hunk 7 above)
Comment (unode) [reply nixos/modules/services/backup/burp.nix:76]: This translates to [systemd OnCalendar time specifications](https://www.freedesktop.org/software/systemd/man/systemd.time.html]. A enum maybe restrict more than desired.

File: nixos/modules/services/backup/burp.nix (hunk 1 above)
Comment (unode) [reply nixos/modules/services/backup/burp.nix:216]: The server but I'm not sure I understand what the question is getting at.

File: nixos/modules/misc/ids.nix (hunk 5 above)
Comment (tokudan) [reply nixos/modules/misc/ids.nix:349]: IIRC dynamic users potentially requires a chown run on start of the service, which is not practical on a big data store with potentially millions of files, so I don't think using dynamic users feasable for burp.

**Comment by unode** [comment]
Addressed the conflict, which is likely to reappear if new services are added.

Working fine on 21.05.

File: nixos/modules/services/backup/burp.nix [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+
+let
+  libDir = "/var/lib/burp";
+
+  cfg = config.services.burp;
+  burp_conf = pkgs.writeText "burp.conf" ''
+    mode = client
+    pidfile = /run/burp/burp.pid
+
+    port = ${toString cfg.client.port}
```
Comment (Infinisil) [inline nixos/modules/services/backup/burp.nix:13]: Already mentioned this in https://github.com/NixOS/nixpkgs/pull/48762#issuecomment-513591625, but by now we should really be using the approach from NixOS/rfcs#42 for this, see also https://github.com/NixOS/nixpkgs/issues/144575. By now there are a lot of NixOS modules which make use of that approach which you can use for reference. Grep for `freeformType` in nixpkgs to see them. You can also check out [the documentation](https://nixos.org/manual/nixos/stable/index.html#sec-freeform-modules).

File: nixos/modules/services/backup/burp.nix [hunk 12]
```diff
@@ -0,0 +1,451 @@
...
+        default = "";
+        description = ''
+          Extra configuration for burp client. Contents will be added verbatim to the
+          configuration file.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.client.enable or cfg.server.enable {
```
Comment (Infinisil) [inline nixos/modules/services/backup/burp.nix:396]: `or` isn't a boolean operator in Nix, check out the [list of operators](https://nixos.org/manual/nix/stable/expressions/language-operators.html). You meant to use `||` here instead. Also needs parenthesis.

**Review by infinisil — CHANGES_REQUESTED** [review]

**Comment by arjan-s** [comment]
Are you still planning to work on this. @unode ? I could really use it. :)

**Comment by unode** [comment]
I'm still using the original recipe and didn't get to implement the suggested approach.
Feel free to push or suggest changes. I don't think I'll be able to get to it anytime soon.

**Pull request closed without merging**